File: CSE241 - Fall 2019/hw6/GTUSet.cpp
#include "GTUSet.h"



namespace kal{
    template <typename T>
    GTUSet<T>::GTUSet(){
        size_ = 1;
        used = 0;
        std::shared_ptr<T> temp(new T[size_], std::default_delete<T[]>());
        this->data = temp;
    }

    template <typename T>
    int GTUSet<T>::max_size() const noexcept{
        return size_;
    }

    template <typename T>
    int GTUSet<T>::size() const noexcept{      
        return used;
    }


    template <typename T>
    bool GTUSet<T>::empty() const noexcept{
        return (used == 0);
    }
    template<typename T>
    void GTUSet<T>::erase(const T& val) {
        int count = 0,index = -1,i = 0;
        std::shared_ptr< T > temp(new T[size_], std::default_delete<T[]>());
        for(auto it = this->begin() ; it != this->end() ; ++it){
            ++count;
            if( (*it) == val){            
                index = count - 1;
            }
            
        }
        if(index == -1) {
         throw invalid_argument("This value is not in the set");
        }
        for(auto it = this->begin() ; it != this->end() ; ++it){
            if( *(it) != this->data.get()[index]){
                temp.get()[i] = *(it);
                ++i;
            }

        }
        this->data = nullptr;
        this->data = temp;
        --used;
    }
    template<typename T>
    void GTUSet<T>::erase(GTUIterator<T> position){
        int index = -1,count = 0;
        for(auto it = this->begin() ; it != this->end() ; ++it){
            ++count;
            if( it == position){            
                index = count - 1;
            }

        }
        if(index == -1){
            cout << "Invalid iterator" << endl;
            return;
        }
        if(index == used - 1) --used;
        else{
            int j = index;    
            for(auto it = position  ; it != this->end() ; ++it){
                data.get()[j] = data.get()[j + 1];
                ++j;
            }
            --used;
        }
    }

    template <typename T>
    void GTUSet<T>::insert(const T& val){
        int i = 0;
        if(size_ <= used || size_ == 0){
            std::shared_ptr<T> temp(new T[size_ * 2], std::default_delete<T[]>());
            for(auto it = this->begin() ; it != this->end() ; ++it){
                if( (*it) == val){
                    throw invalid_argument("This value is already in the set!");
                }
                temp.get()[i] = *(it);
                ++i;
            }
            temp.get()[i] = val;
             this->data = nullptr;
             this->data = temp;           
             size_ = size_ * 2;
             ++used;
        }
        else{
            for(auto it = this->begin() ; it != this->end() ; ++it){
                if( (*it) == val){
                    return;
                }
            }        
                data.get()[used] = val;
                ++used;              

        }

        
    }

    template<typename T>
    void GTUSet<T>::clear() noexcept{
        this->used = 0;
        this->data = nullptr;
    }




}

File: CSE241 - Fall 2019/hw8/src/GTUSet.java
import java.lang.reflect.Array;
import java.security.InvalidParameterException;

public class GTUSet<T> extends GTUContainer<T>{

    public GTUSet(Class <T[]> type){
        type_ = type;
        capacity = 1;
        used = 0;
        data = type.cast(Array.newInstance(type.getComponentType(),capacity));
    }

    public class GTUSetIterator implements GTUIterator<T>{
        int pos = 0;

        public T next(){
            return data[pos++];
        }

        public boolean hasNext(){
            if(pos < used)
                return true;
            
            return false;
        }
    }
    @Override
    public boolean empty(){
        return (used == 0);
    }
    @Override
    public void clear(){
        used = 0;
        data = null;
    }
    @Override
    public int max_size(){
        return capacity;
    }

    @Override
    public int size(){
        return used;
    }

    @Override
    public void erase(Object obj){
        boolean check = false;
        int index = -1;
        for(int i = 0 ; i < used ; ++i){
            if(data[i] == obj){
                check = true;
                index = i;
            }
        }

        if(check){
            for(int i = index ; i < used - 1; ++i){
                data[i] = data[i + 1];
            }
            --used;

        }
    }

    @Override
    public void insert(T obj) throws InvalidParameterException{
        if(capacity <= used || capacity == 0){
            T[] temp = type_.cast(Array.newInstance(type_.getComponentType(), capacity * 2));
            for(int i = 0 ; i < used ; ++i){
                if(data[i].equals(obj)){
                    throw new InvalidParameterException("Element already exists.");
                }
                temp[i] = data[i];         
            }
            data = null;
            data = type_.cast(Array.newInstance(type_.getComponentType(), capacity * 2));
            capacity *= 2;
            for(int i = 0 ; i < used ; ++i)
                data[i] = temp[i];
            data[used] = obj;
            ++used;
        } 
        else{
            data[used] = obj;
            ++used;
        }
    }

    @Override
    public boolean contains(Object o){

        for(int i = 0 ; i < used ; ++i){
            if(data[i] == o)
                return true;
        }
        return false;
    }

    public GTUIterator<T> iterator(){
        return new GTUSetIterator();
    }


}